content/docs/iac/guides/building-extending/providers/implementers/protocol-reference.md

```diff
@@ -0,0 +1,347 @@
+---
+title_tag: "Provider protocol reference"
+meta_desc: "Complete reference for the Pulumi provider gRPC protocol, including all RPC methods and their semantics."
+title: Provider protocol reference
+h1: Provider protocol reference
+meta_image: /images/docs/meta-images/docs-meta.png
+menu:
+    iac:
+        name: Protocol reference
+        parent: iac-guides-provider-implementers
+        weight: 10
+---
+
+This document describes the gRPC protocol that all Pulumi providers implement. Understanding this protocol is essential for [implementing providers directly](/docs/iac/guides/building-extending/providers/provider-architecture/#layer-2-generated-language-bindings) without a higher-level SDK.
+
+The complete protocol definition is in [`provider.proto`](https://github.com/pulumi/pulumi/blob/master/proto/pulumi/provider.proto). You can also find auto-generated documentation at [pulumi.io/docs/proto](https://pulumi.io/docs/proto/).
+
+## Overview
+
+Pulumi providers are gRPC servers. When a Pulumi program runs, the engine starts provider processes and communicates with them over gRPC. The provider implements the `ResourceProvider` service, which defines methods for managing resources throughout their lifecycle.
+
+### How it works
+
+The Pulumi engine starts your provider as a subprocess and connects to it via gRPC. It then calls `Configure` with credentials and settings to initialize the provider. For each resource operation, the engine calls the appropriate method such as `Check`, `Diff`, or `Create`. Your provider executes the operation and returns results, while the engine manages state and orchestrates the overall deployment.
+
+## Essential methods
+
+These methods are required for a functional provider:
+
+### GetSchema
+
+Returns the provider's schema, which defines all resources, their properties, and configuration options.
+
+```protobuf
+rpc GetSchema(GetSchemaRequest) returns (GetSchemaResponse)
+```
+
+**Request fields:**
+
+- `version` (int32): Schema version to retrieve (use 0 for current)
+- `subpackage_name` (string): Optional subpackage name
+- `subpackage_version` (string): Optional subpackage version
+
+**Response fields:**
+
+- `schema` (string): JSON-encoded [package schema](/docs/iac/guides/building-extending/packages/schema/)
+
+The schema is critical—it tells Pulumi what resources your provider offers, what inputs they accept, and what outputs they produce. Without a valid schema, Pulumi cannot generate SDKs or validate programs.
+
+### Configure
+
+Initializes the provider with configuration values like credentials and region settings.
+
+```protobuf
+rpc Configure(ConfigureRequest) returns (ConfigureResponse)
+```
+
+**Request fields:**
+
+- `variables` (map<string, string>): Environment variables
```

Comment:
This isn't environment variables, it's the same as "args" but string-ified, we don't recommend using it anymore.